src/AI/AICombatUtils.java

```diff
@@ -187,7 +198,7 @@ public static int findMaxMobileWeaponRange(Commander co)
       final XYCoord attackerCoord = new XYCoord(u);
       boolean requiresMoving = false;
       int dist = targetCoord.getDistance(attackerCoord);
-      if( !u.canAttack(target.model, dist, requiresMoving) )
+      if( !u.canAttack(gameMap, target.model, dist, requiresMoving) )
```

Comment:
We didn't need a `gameMap` here before. I don't think we really do now, either, except that `UnitContext` requires it. I don't really like how `UnitContext` is trying to be all things to all people, or the fact that we now have functions like `getRangeContext`. That smells to me, probably because it means we don't really know what we have if we have a `UnitContext` object, unless we just purpose-built it; I'd be vaguely curious how often we create them.

At any rate, it definitely feels like there should be a simpler and cleaner way to handle this.

Reply:
> We didn't need a `gameMap` here before. I don't think we really do now, either, except that `UnitContext` requires it.

`UnitContext` does not require it; however, I prefer to provide maximum information to make the results as accurate as possible. For example, not having the map means we won't have knowledge of the weather, which means we won't apply -1 range if there's a sandstorm going.
> I don't really like how `UnitContext` is trying to be all things to all people

I had the same complaint about the `GameInputState` state data, though that's a much smaller code surface area, to be fair.
The core issue here is that `UnitContext` is basically the information-carrier to enable all the extension points that `Commander` wants. We *could* make a bunch of classes with subsets of `UnitContext`'s current fields for calculating different things, but we'd be likely to end up with even more duplication than we already have.
>, or the fact that we now have functions like `getRangeContext`.

`getRangeContext()` exists because I didn't want to calculate min and max range separately (I don't feel strongly about this; maybe we should?), and didn't feel like making up an entirely new object for the purpose.
> That smells to me, probably because it means we don't really know what we have if we have a `UnitContext` object, unless we just purpose-built it;

You are indeed correct. A `UnitContext` is modified based on... context. I don't need to touch the range if all I'm calculating is movement, for example. ...At least until someone makes a CO that changes stuff based on other unit stats.
We could just institute a policy of "apply all modifications to every facet on instantiation", but that would require some rewickering of stuff.
> I'd be vaguely curious how often we create them.

Depends on use case, but "a lot".
For example, 4 minimum for combat calculations, usually more.
> At any rate, it definitely feels like there should be a simpler and cleaner way to handle this.

I'm trying to design this to support a lot of features we don't have yet, so new COs and such can just slot in without changing anything outside the new code and code to enable invoking the new code.
That necessarily involves pulling in a lot of data.
If there is a cleaner solution, I'm not sure what it would look like. Maybe that immutable state thing you were talking about before?

Reply:
Not super thrilled with creating a million objects all the time - seems like it would create a lot of extra runtime work that we didn't use to have. I also don't know that I want to enforce "full" population of `UnitContext` every time one is created, since that just creates more busywork for no apparent gain.
The immutable state thing would be fun to try, but would also require quite a lot of rewickering.

Reply:
> Not super thrilled with creating a million objects all the time - seems like it would create a lot of extra runtime work that we didn't use to have.

Technically correct, but the only time the actual game engine takes up most of our processing is when the AI is active.
If it ends up eating lots of our CPU budget, then I'd expect it to be a case of e.g. running a calculation every frame when we don't need to.
As for creating lots of objects... that's kinda the nature of the beast. You need scratch space to calculate stuff.
> I also don't know that I want to enforce "full" population of `UnitContext` every time one is created, since that just creates more busywork for no apparent gain.

Being able to not populate all of its fields is a core feature. I wouldn't want to drop that.
What I was proposing was to apply all *modifications* to the `UnitContext` on instantiation, so you don't ever have to wonder whether e.g. the context's range has been modified yet.
It would certainly be more CPU work, and would require rewickering the combat stuff a bit (since that operates on the assumption that it gets a "clean" context that it can then clone and modify in various ways)

Reply:
> As for creating lots of objects... that's kinda the nature of the beast. You need scratch space to calculate stuff.

We didn't do that before, and I hope you aren't arguing that this is the only valid solution. Things like this are probably why OO has a bit of a reputation for inefficiency and bloat. Once you start thinking in objects, everything becomes one.

> What I was proposing was to apply all modifications

Would that even make sense? E.g. if we only partly populate the context, and then apply a modifier that affects something unpopulated? At any rate, I'm not sure that's the right answer either.

Reply:
> We didn't do that before, and I hope you aren't arguing that this is the only valid solution.

There are certainly more possible solutions, but I don't see any with better tradeoffs than this one.
> Things like this are probably why OO has a bit of a reputation for inefficiency and bloat.

Absolutely correct, but I don't feel like it matters that much in this case.
> Would that even make sense? E.g. if we only partly populate the context, and then apply a modifier that affects something unpopulated?

Sure. I want to know about movement, so I don't populate the weapon. I apply range modifications, and those range modifications see that there's no weapon, so they no-op (or change a meaningless range number)